### PR TITLE
board_ctrl: Clean up usage of VBUS, nARMED from user/kernelspace

### PR DESCRIPTION
**Pull request by pussuw**
Move logic implemented in the header files to source files, this way
it will be simpler to define which is compiled to kernel space and
which to user space

Also allows to remove some headers that pull in half the universe
from the board definitions (which should just be pin definitions and
no functionality)

### CODE REVIEW

File: platforms/common/board_common.c [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+
+/****************************************************************************
+ * Public Functions
+ ****************************************************************************/
+
+#if defined(GPIO_OTGFS_VBUS) && \
+    defined(CONFIG_BUILD_FLAT) || !defined(__PX4_NUTTX)
+
+/* Default implementation for POSIX and flat NUTTX if the VBUS pin exists */
+int board_read_VBUS_state(void)
```
Comment (pussuw) [inline platforms/common/board_common.c:50]: FAILED: platforms/common/CMakeFiles/px4_platform.dir/board_common.c.o 

../../platforms/common/board_common.c: In function ‘board_read_VBUS_state’:
../../platforms/common/board_common.c:52:27: error: ‘GPIO_OTGFS_VBUS’ undeclared (first use in this function)
   52 |  return px4_arch_gpioread(GPIO_OTGFS_VBUS);
      |                           ^~~~~~~~~~~~~~~

OK, So what am I missing here ?

File: platforms/common/board_common.c [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+#include "board_config.h"
+
+/****************************************************************************
+ * Public Functions
+ ****************************************************************************/
+
+#if defined(GPIO_OTGFS_VBUS) && \
+    (defined(CONFIG_BUILD_FLAT) || !defined(__PX4_NUTTX))
+
+/* Default implementation for POSIX and flat NUTTX if the VBUS pin exists */
```
Comment (jlaitine) [inline platforms/common/board_common.c:49]: Can the GPIO_OTGFS_VBUS actually exist in POSIX build?

File: platforms/nuttx/src/px4/common/board_ctrl.c [hunk 3]
```diff
@@ -1,6 +1,6 @@
 /****************************************************************************
  *
- *   Copyright (C) 2020 Technology Innovation Institute. All rights reserved.
+ *   Copyright (C) 2022 Technology Innovation Institute. All rights reserved.
```
Comment (jlaitine) [inline platforms/nuttx/src/px4/common/board_ctrl.c:3]: I don't think it is necessary to update the year

File: platforms/nuttx/src/px4/common/board_ctrl.c [hunk 4]
```diff
@@ -31,160 +31,44 @@
...
-	}
-
-	if (builtin) {
-		/* This is running in the userspace thread, created by nsh, and
-		   called via boardctl. Call the main directly */
-		return builtin->main(argc, argv);
-	}
-
-	return ENOENT;
+#if defined GPIO_nARMED
```
Comment (jlaitine) [inline platforms/nuttx/src/px4/common/board_ctrl.c:60]: #if defined(GPIO_nARMED)

File: platforms/nuttx/src/px4/common/include/px4_platform/board_ctrl.h [hunk 5]
```diff
@@ -46,11 +46,11 @@
 #define IOCTL_IDX_TO_BASE(x)	((((x) & 0xF) << 8) | _BOARDCTLIOCBASE)
 #define IOCTL_BASE_TO_IDX(x)	(((x) & 0x0F00) >> 8)
 
-#define _ORBIOCDEVBASE				IOCTL_IDX_TO_BASE(0)
+#define _ORBIOCDEVBASE			IOCTL_IDX_TO_BASE(0)
 #define _HRTIOCBASE				IOCTL_IDX_TO_BASE(1)
```
Comment (jlaitine) [inline platforms/nuttx/src/px4/common/include/px4_platform/board_ctrl.h:50]: while at it, maybe indent this as well :)

File: boards/px4/fmu-v5/src/board_config.h [hunk 6]
```diff
@@ -226,24 +222,12 @@
 #define GPIO_nARMED          /* PI0 */  (GPIO_OUTPUT|GPIO_PUSHPULL|GPIO_SPEED_2MHz|GPIO_OUTPUT_CLEAR|GPIO_PORTI|GPIO_PIN0)
 
 /* For protected build, define the LOCKOUT_STATE macros as function calls */
-#ifdef CONFIG_BUILD_FLAT
+#if defined(CONFIG_BUILD_FLAT)
 #define BOARD_INDICATE_EXTERNAL_LOCKOUT_STATE(enabled)  px4_arch_configgpio((enabled) ? GPIO_nARMED : GPIO_nARMED_INIT)
-#define BOARD_GET_EXTERNAL_LOCKOUT_STATE() px4_arch_gpioread(GPIO_nARMED)
+#define BOARD_GET_EXTERNAL_LOCKOUT_STATE() 				px4_arch_gpioread(GPIO_nARMED)
```
Comment (jlaitine) [inline boards/px4/fmu-v5/src/board_config.h:227]: unnecessary change

File: boards/px4/fmu-v5/src/board_config.h [hunk 7]
```diff
@@ -226,24 +222,12 @@
...
-static inline bool board_get_external_lockout_state(void)
-{
-	platformioclockoutstate_t state = {false};
-	boardctl(PLATFORMIOCGETLOCKOUT, (uintptr_t)&state);
-	return state.enabled;
-}
-#define BOARD_INDICATE_EXTERNAL_LOCKOUT_STATE(enabled) board_indicate_external_lockout_state(enabled)
-#define BOARD_GET_EXTERNAL_LOCKOUT_STATE() board_get_external_lockout_state()
+#define BOARD_INDICATE_EXTERNAL_LOCKOUT_STATE(enabled) 	boardctrl_indicate_external_lockout_state(enabled)
+#define BOARD_GET_EXTERNAL_LOCKOUT_STATE() 				boardctrl_get_external_lockout_state()
```
Comment (jlaitine) [inline boards/px4/fmu-v5/src/board_config.h:230]: also here no need to extend the line lenght with n tabs...?

File: platforms/common/board_common.c (hunk 2 above)
Comment (pussuw) [reply platforms/common/board_common.c:49]: I'm not sure. The original implementation in board_common.h was:

```
#if defined(GPIO_OTGFS_VBUS)
#  define board_read_VBUS_state() (px4_arch_gpioread(GPIO_OTGFS_VBUS) ? 0 : 1)
#else
int board_read_VBUS_state(void);
#endif
```
So its non-existence in POSIX is implicit, which I cannot know

File: platforms/nuttx/src/px4/common/board_ctrl.c (hunk 4 above)
Comment (pussuw) [reply platforms/nuttx/src/px4/common/board_ctrl.c:60]: Yes

**Comment by jlaitine** [comment]
Looks nice, added a few style comments and one question

File: boards/px4/fmu-v5/src/board_config.h (hunk 6 above)
Comment (pussuw) [reply boards/px4/fmu-v5/src/board_config.h:227]: Editor removed the trailing TAB:s automatically, I can add them back if necessary

Comment (pussuw) [reply boards/px4/fmu-v5/src/board_config.h:227]: Rest of the file tries to indent to the same TAB boundary. I think there are mixed single whitespace + TAB characters here now.

File: boards/px4/fmu-v5/src/board_config.h (hunk 7 above)
Comment (pussuw) [reply boards/px4/fmu-v5/src/board_config.h:230]: Rest of the file tries to indent to the same TAB boundary. I think there are mixed single whitespace + TAB characters here now.

File: platforms/nuttx/src/px4/common/include/px4_platform/board_ctrl.h (hunk 5 above)
Comment (pussuw) [reply platforms/nuttx/src/px4/common/include/px4_platform/board_ctrl.h:50]: Hmm, on my editor they are indented to the same TAB boundary, maybe there are mixed single whitespace + TABs here too

**Comment by pussuw** [comment]
px4/fmuv5:protected broke, the autobuild does not detect this. The new patch fixes (hopefully) the build issue.

**Review by jlaitine — APPROVED** [review]
LGTM!

Do you think that this could maybe be upstreamed before merging? If there are requests for style or such, we'd avoid dome merge conflicts in the future.

**Comment by pussuw** [comment]
> LGTM!
> 
> Do you think that this could maybe be upstreamed before merging? If there are requests for style or such, we'd avoid dome merge conflicts in the future.

Can be done, I'll do upstreaming for all my changes for mpfs protected mode in a bunch. 1-2 are still coming. I'll apply the chainsaw when updating / backporting our fork from upstream
